### ISSUE TITLE
Change Span name after starting a span.

### ISSUE THREAD

**Issue by g-easy** [census-instrumentation/opencensus-specs#217]
Currently, the Span name can be set at start time, but (in Java and C++) can't be changed afterwards.

opencensus-go has [Span.SetName](https://github.com/census-instrumentation/opencensus-go/blob/aab39bd6a98b853ab66c8a564f5d6cfcad59ce8a/trace/trace.go#L298).

Others have this functionality, e.g. Zipkin.

Envoy uses this functionality to sometimes change the span name if a "decorator" is configured.

We should decide whether Span names can be changed after starting a span. And also how this will interact with z-pages.

**Comment by adriancole** [comment]
yep in a lot of our instrumentation, names are resetable until late.
This is because often a better span name isn't available, yet. For
example, the route name of an http request is usually not available
until late in processing, sometimes not until the response.

**Comment by SergeyKanzhelev** [comment]
The only question should sampling be re-run as sampler today receives a span name as an argument. So potentially a name-based sampling decision may have already been done or worse - other decision needs to be done based on the proper name like `route`

**Comment by g-easy** [comment]
Enabling sampling after starting a span is mentioned in #218 also.

A potential problem I can see is: if sampling_enabled is stored in the SpanContext, and the SpanContext has been copied, there isn't a reasonable way of updating all the copies with the new sampling decision.

Maybe a better approach would be to start a new span after the routing decision is made.

**Comment by SergeyKanzhelev** [comment]
@g-easy the problem with two spans is a bad query and analysis experience. Basically two spans will only differ in most cases in a few milliseconds of duration and a name. We can just call this potential problem an edge case and ignore it.

**Comment by g-easy** [comment]
On the flip side: would it be a better approach to delay starting the span until the name is known?
